File: python/hash-table/copy-list-with-random-pointer.py
# Definition for singly-linked list with a random pointer.
# class RandomListNode(object):
#     def __init__(self, x):
#         self.label = x
#         self.next = None
#         self.random = None

class Solution(object):
    def copyRandomList(self, head):
        """
        :type head: RandomListNode
        :rtype: RandomListNode
        """
        root = RandomListNode(0)
        root.next = head
        while head:
            new_node = RandomListNode(head.label)
            new_node.next = head.next
            head.next = new_node
            head = head.next.next

        head = root.next
        while head:
            new_node = head.next
            if head.random:
                new_node.random = head.random.next

            head = head.next.next


        prev = root
        head = root.next
        while head:
            new_node = head.next
            head.next = head.next.next
            prev.next = new_node
            prev = new_node
            head = head.next

        return root.next

File: c/hash-table/copy-list-with-random-pointer.c
/**
 * Definition for singly-linked list with a random pointer.
 * struct RandomListNode {
 *     int label;
 *     struct RandomListNode *next;
 *     struct RandomListNode *random;
 * };
 */
struct RandomListNode *copyRandomList(struct RandomListNode *head) {
    struct RandomListNode root;
    struct RandomListNode *pre;
    struct RandomListNode *new_node;

    root.next = head;
    while(head){
        new_node = malloc(sizeof(struct RandomListNode));
        new_node->next = head->next;
        new_node->label = head->label;
        head->next = new_node;
        head = head->next->next;
    }

    head = root.next;
    while(head){
        new_node = head->next;
        if(head->random)
            new_node->random = head->random->next;
        else
            new_node->random = NULL;

        head = head->next->next;
    }

    head = root.next;
    pre = &root;
    while(head){
        new_node = head->next;
        head->next = new_node->next;
        head = head->next;

        pre->next = new_node;
        pre = new_node;
    }

    pre->next = NULL;

    return root.next;
}
